### DISCUSSION TITLE
Filtering data using binary attributes

### DISCUSSION THREAD

**Discussion by apalmer0** [visgl/deck.gl#6953] [Q&A]
I'm having trouble filtering my TripsLayer, to which I'm supplying binary attributes for the data. As it says in the docs for accessor functions (in this case, `getFilterValue`), since the data is a non-iterable object, the current element in the data stream is always null and I'm not able to selectively filter out specific elements. 

Based on this, I would assume it's not possible to filter layers when using binary attributes, but I also have a GeoJsonLayer that also uses binary attributes (we've implemented our own `geojsonToBinary` utility), and that one _is_ filterable - the `getFilterValue` function _does_ receive the current element in the data stream. 

My TripsLayer data looks like this:

```js
// TripsLayer

{
  length: data.length,
  startIndices: new Uint32Array(startIndices),
  properties: data,
  attributes: {
    getPath: {
      value: new Float64Array(positions),
      size: 3,
    },
    getTimestamps: {
      value: new Float32Array(timestamps),
      size: 1,
    },
  },
}
```

and my GeoJsonLayer data looks like this:

```js
// GeoJsonLayer

lines: {
  type: 'LineString',
  attributes: {
    getColor: { value: colors, size: 4 },
    getWidth: { value: new Uint16Array(data.width), size: 1 },
  },
  positions: { value: positions, size: 3 },
  pathIndices: { value: new Uint32Array(data.pathIndices), size: 1 },
  globalFeatureIds: { value: featureIds, size: 1 },
  featureIds: { value: featureIds, size: 1 },
  numericProps: {},
  properties,
},
points: {
  type: 'Point',
  positions: { value: new Float32Array([]), size: 3 },
  globalFeatureIds: { value: new Uint16Array([]), size: 1 },
  featureIds: { value: new Uint16Array([]), size: 1 },
  numericProps: {},
  properties: [],
},
polygons: {
  type: 'Polygon',
  positions: { value: new Float32Array([]), size: 3 },
  polygonIndices: { value: new Uint16Array([]), size: 1 },
  primitivePolygonIndices: { value: new Uint16Array([]), size: 1 },
  triangles: { value: [], size: 1 },
  globalFeatureIds: { value: new Uint16Array([]), size: 1 },
  featureIds: { value: new Uint16Array([]), size: 1 },
  numericProps: {},
  properties: [],
},
```

Is there a property missing from my TripsLayer data object that's provided by the data in my GeoJsonLayer? Or is there some inherent difference between these two layer types that allows the GeoJsonLayer to handle non-iterable data differently than the TripsLayer? Or is there something else I might be doing wrong?

Apologies if this isn't enough information to go on... happy to add any more details that'll help. Thanks in advance!

**Comment by Pessimistress** [comment]
> since the data is a non-iterable object, the current element in the data stream is always null and I'm not able to selectively filter out specific elements.

The accessors always receive two arguments. When you use non-iterable data you need the second argument to figure out which element is being evaluated, see https://deck.gl/docs/developer-guide/using-layers#accessors

You can also supply a binary attribute to `getFilterValue` the same way you do path and timestamps.
